main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -514,53 +501,66 @@ object Safety {
         acc + (signature -> m)
     }
   }
+  
+  /**
+    * Convert a `java.lang.reflect.Method` to a MethodSignature.
+    */
+  private def getJavaMethodSignature(method: java.lang.reflect.Method) = {
+    MethodSignature(method.getName(), 
+      getFlixType(method.getReturnType),
+      method.getParameterTypes().toList.map(getFlixType))
+  }
+
+  /**
+    * Given a class, get sets representing both the methods that can be implemented,
+    * and the methods that must be implemented.
+    */
+  private def getJavaMethods(clazz: java.lang.Class[_]): (Set[MethodSignature], Set[MethodSignature]) = {
+    val methods = clazz.getMethods().toList.filterNot(m => java.lang.reflect.Modifier.isStatic(m.getModifiers()))
+    val mustImplement = if (clazz.isInterface()) {
+        methods.filterNot(_.isDefault())
+      } else {
+        methods.filter(m => java.lang.reflect.Modifier.isAbstract(m.getModifiers()))
+      }
+    (methods.map(getJavaMethodSignature).toSet, mustImplement.map(getJavaMethodSignature).toSet)
+  }
 
   /**
-    * Ensures that `clazz` is a Java interface, and that `methods` fully implement it.
+    * Ensures that `methods` fully implement `clazz`
     */
   private def checkObjectImplementation(clazz: java.lang.Class[_], tpe: Type, methods: List[JvmMethod], loc: SourceLocation): List[CompilationMessage] = {
     // 
-    // Check that `clazz` is an interface
+    // Check that the first argument looks like "this"
     // 
-    if (!clazz.isInterface()) {
-      List(IllegalObjectDerivation(clazz, loc))
-    } else {
-      // 
-      // Check that the first argument looks like "this"
-      // 
-      val thisErrors = methods.flatMap {
-        case JvmMethod(ident, fparams, _, _, _, _, methodLoc) => 
-          if (fparams.length < 1)
-            Some(MissingThis(tpe, ident.name, methodLoc))
-          else if (fparams.head.tpe != tpe) {
-            Some(IllegalThisType(tpe, fparams.head.tpe, ident.name, methodLoc))
-          } else {
-            None
-          }
-      }
+    val thisErrors = methods.flatMap {
+      case JvmMethod(ident, fparams, _, _, _, _, methodLoc) => 
+        if (fparams.length < 1)
+          Some(MissingThis(tpe, ident.name, methodLoc))
+        else if (fparams.head.tpe != tpe) {
```

Comment:
This check is not exactly correct, but I guess that will be discovered/refactored later.

Reply:
Can you give me a hint as to what's not correct? 🤣

Reply:
Yes, but its complicated. `==` only checks for structural equality, but there are types which are equivalent but not structurally equivalent. For example the function type `Int32 -> Int32 \ {r1, r2}` != `Int32 -> Int32 \ {r2, r1}`. But, I don't think this will start to hurt until much later :)

Reply:
Actually, maybe this check should happen in the Typer?

Reply:
Well, I originally suggested doing this in the typer, but you suggested I shouldn't 🤣

Remember that the types here will only ever be Java types. So `Int32 -> Int32 \ {r1, r2}` (and similar) will never be relevant?

Reply:
Possibly, not sure. But I think its fine for now